### PR TITLE
feat: version compatibility check for MonoVertex

### PR DESCRIPTION
**Pull request by kohlisid**
fixes https://github.com/numaproj/numaflow/issues/1908

### CODE REVIEW

File: serving/source-sink/src/version.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+impl VersionInfo {
+    /// Initializes VersionInfo by reading environment variables or default values
+    fn init() -> Self {
+        let version = env::var("CARGO_PKG_VERSION").unwrap_or_else(|_| "latest".to_string());
+        let build_date =
+            env::var("BUILD_DATE").unwrap_or_else(|_| "1970-01-01T00:00:00Z".to_string());
+        let git_commit = env::var("GIT_COMMIT").unwrap_or_else(|_| "".to_string());
+        let git_tag = env::var("GIT_TAG").unwrap_or_else(|_| "".to_string());
+        let git_tree_state = env::var("GIT_TREE_STATE").unwrap_or_else(|_| "".to_string());
```
Comment (yhl25) [inline serving/source-sink/src/version.rs:46]:
```suggestion
        let git_commit = env::var("GIT_COMMIT").unwrap_or_default();
        let git_tag = env::var("GIT_TAG").unwrap_or_default();
        let git_tree_state = env::var("GIT_TREE_STATE").unwrap_or_default();
```

File: serving/source-sink/src/version.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+pub struct VersionInfo {
+    pub version: String,
+    pub build_date: String,
+    pub git_commit: String,
+    pub git_tag: String,
+    pub git_tree_state: String,
+    pub rust_version: String,
+    pub compiler: String,
+    pub platform: String,
+}
```
Comment (yhl25) [inline serving/source-sink/src/version.rs:36]: Where are we using these fields?

File: serving/source-sink/src/version.rs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    static ref MINIMUM_SUPPORTED_SDK_VERSIONS: SdkConstraints = {
+        let mut m = HashMap::new();
+        m.insert("go".to_string(), "0.7.0-rc2".to_string());
+        m.insert("python".to_string(), "0.7.0a1".to_string());
+        m.insert("java".to_string(), "0.7.2-0".to_string());
+        m.insert("rust".to_string(), "0.0.1".to_string());
+        m
+    };
+}
+
```
Comment (yhl25) [inline serving/source-sink/src/version.rs:24]: use one cell? https://github.com/rust-lang-nursery/lazy-static.rs/issues/214

File: serving/source-sink/src/server_info.rs [hunk 4]
```diff
@@ -0,0 +1,625 @@
+use pep440_rs::{Version as PepVersion, VersionSpecifier};
+use semver::{Version, VersionReq};
+use serde::{Deserialize, Deserializer, Serialize};
+use std::collections::HashMap;
```
Comment (yhl25) [inline serving/source-sink/src/server_info.rs:4]: why do we need `Deserializer`?

File: serving/source-sink/src/server_info.rs [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+    // Extract the server info from the Result
+    let server_info = match read_result {
+        Ok(info) => info,
+        Err(e) => {
+            return Err(Error::ServerInfoError(format!(
+                "Failed to read server info: {}",
+                e
+            )));
+        }
+    };
```
Comment (yhl25) [inline serving/source-sink/src/server_info.rs:63]:
```suggestion
    let server_info = read_server_info(file_path).await?;
```

File: serving/source-sink/src/server_info.rs [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
+    if min_numaflow_version.is_empty() {
+        warn!("Failed to get the minimum numaflow version, skipping numaflow version compatibility check");
+    } else if !numaflow_version.contains("latest")
+        && !numaflow_version.contains(&version_info.git_commit)
+    {
+        // Check the compatibility between the SDK and Numaflow versions
+        // If any error occurs, return the error
+        if let Err(e) = check_numaflow_compatibility(numaflow_version, min_numaflow_version) {
+            return Err(e);
+        }
```
Comment (yhl25) [inline serving/source-sink/src/server_info.rs:88]:
```suggestion
        check_numaflow_compatibility(numaflow_version, min_numaflow_version)?;
```

File: serving/source-sink/src/server_info.rs [hunk 7]
```diff
@@ -0,0 +1,625 @@
...
+
+/// Checks if the given version meets the specified constraint.
+fn check_constraint(version: &Version, constraint: &str) -> error::Result<()> {
+    // Parse the given constraint as a semantic version requirement
+    let version_req = VersionReq::parse(constraint).map_err(|e| {
+        Error::ServerInfoError(format!(
+            "Error parsing constraint: {},\
+         constraint string: {}",
+            e, constraint
+        ))
```
Comment (yhl25) [inline serving/source-sink/src/server_info.rs:112]: why is it split into two lines? fmt?

File: serving/source-sink/src/server_info.rs [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+    // Check if the numaflow version satisfies the constraint
+    if let Err(e) = check_constraint(&numaflow_version_semver, &numaflow_constraint) {
+        let err_string = format!(
+            "numaflow version {} must be upgraded to at least {}, in order to work with current SDK version {}",
+            numaflow_version_semver, min_numaflow_version, e
+        );
+        return Err(Error::ServerInfoError(err_string));
+    }
+    Ok(())
+}
```
Comment (yhl25) [inline serving/source-sink/src/server_info.rs:148]: use map err?

File: serving/source-sink/src/server_info.rs [hunk 9]
```diff
@@ -0,0 +1,625 @@
...
+                Error::ServerInfoError(format!("Error parsing SDK constraint: {}", e))
+            })?;
+
+            if !specifiers.contains(&sdk_version_pep440) {
+                let err_string = format!(
+                    "SDK version {} must be upgraded to at least {}, in order to work with the current numaflow version",
+                    sdk_version_pep440, sdk_required_version
+                );
+                return Err(Error::ServerInfoError(err_string));
+            }
```
Comment (yhl25) [inline serving/source-sink/src/server_info.rs:175]:
```suggestion
            if !specifiers.contains(&sdk_version_pep440) {
                return Err(Error::ServerInfoError(format!(
                    "SDK version {} must be upgraded to at least {}, in order to work with the current numaflow version",
                    sdk_version_pep440, sdk_required_version
                )));
            }
```

File: serving/source-sink/src/server_info.rs [hunk 10]
```diff
@@ -0,0 +1,625 @@
...
+        } else {
+            // Strip the 'v' prefix if present for non-Python languages
+            let sdk_version_stripped = if sdk_version.starts_with('v') {
+                &sdk_version[1..]
+            } else {
+                sdk_version
+            };
```
Comment (yhl25) [inline serving/source-sink/src/server_info.rs:182]:
```suggestion
            // Strip the 'v' prefix if present for non-Python languages
            let sdk_version_stripped = sdk_version.trim_start_matches('v');
```

File: serving/source-sink/src/server_info.rs [hunk 11]
```diff
@@ -0,0 +1,625 @@
...
+
+            // Check if the SDK version satisfies the constraint
+            if let Err(e) = check_constraint(&sdk_version_semver, &sdk_constraint) {
+                let err_string = format!(
+                    "SDK version {} must be upgraded to at least {}, in order to work with the current numaflow version: {}",
+                    sdk_version_semver, sdk_required_version, e
+                );
+                return Err(Error::ServerInfoError(err_string));
+            }
+        }
```
Comment (yhl25) [inline serving/source-sink/src/server_info.rs:196]:
```suggestion
            check_constraint(&sdk_version_semver, &sdk_constraint).map_err(|_| {
                Error::ServerInfoError(format!(
                    "SDK version {} must be upgraded to at least {}, in order to work with the current numaflow version",
                    sdk_version_semver, sdk_required_version
                ))
            })?;
```

File: serving/source-sink/src/server_info.rs [hunk 12]
```diff
@@ -0,0 +1,625 @@
...
+        retry += 1;
+        if retry >= 10 {
+            // Return an error if the retry limit is reached
+            return Err(Error::ServerInfoError(
+                "server-info reading retry exceeded".to_string(),
+            ));
+        }
+        sleep(Duration::from_millis(100)).await; // Sleep before retrying
+    }
+
```
Comment (yhl25) [inline serving/source-sink/src/server_info.rs:244]: use our retry?

File: serving/source-sink/src/lib.rs [hunk 13]
```diff
@@ -33,6 +34,11 @@ pub mod transformer;
 pub mod forwarder;
 
 pub mod message;
+
+pub mod server_info;
+
+pub mod version;
```
Comment (vigith) [inline serving/source-sink/src/lib.rs:40]: these need not be `pub` also `mod version` should be inside `server_info`

**Pull request closed without merging**